### ISSUE TITLE
Incompatibility issue when install

### ISSUE THREAD

**Issue by IsmailAhmedIsmail** [Neo4j-Swift/Neo4j-Swift#1]
I'm trying to add Neo4j-Swift as a dependency in my vapor project.
Here is my Packages.swift file
<img width="785" alt="Screen Shot 2019-08-24 at 3 40 50 PM" src="https://user-images.githubusercontent.com/17710260/63638230-d354a280-c685-11e9-8acf-9412b9d23c2d.png">

However when I run swift build in terminal I get the following error:

error: the package neo4j-swift[https://github.com/Neo4j-Swift/Neo4j-Swift.git] @ 5.0.0 contains incompatible dependencies:
    bolt-swift[https://github.com/Neo4j-Swift/Bolt-swift.git] @ master

**Comment by teameh** [comment]
I had the same problem. Fixed this by adding both Bolt and Neo4j's master version to the dependenies as well. Also had to use vapor beta. 

```
dependencies: [
        .package(url: "https://github.com/vapor/vapor.git", from: "4.0.0-beta.1"),
        .package(url: "https://github.com/Neo4j-Swift/Neo4j-Swift.git", .branch("master")),
        .package(url: "https://github.com/Neo4j-Swift/Bolt-swift.git", .branch("master")),
]
```

**Comment by niklassaers** [comment]
Ah, sorry I hadn't replied earlier. Current Bolt-Swift uses SwiftNIO 2, while Vapor 3 users SwiftNIO 1. Having both in the same project doesn't work, so either having Vapor 4 like @teameh does or specifying a previous version of Bolt should fix it.

**Comment by teameh** [comment]
Thanks for the confirmation. However, I did not manage to get it working. 

I've got everything builded with the following dependencies config:

    dependencies: [
        .package(url: "https://github.com/vapor/vapor.git", from: "4.0.0-beta"),
        .package(url: "https://github.com/vapor/fluent.git", from: "4.0.0-beta"),
        .package(url: "https://github.com/vapor/fluent-sqlite-driver.git", from: "4.0.0-beta"),
        .package(url: "https://github.com/Neo4j-Swift/Neo4j-Swift.git", .branch("master")),
    ],

And trying to connect with:

    func routes(_ app: Application) throws {
        app.get { req -> String in
            print("get /")
            class CustomConfig: ClientConfigurationProtocol {
                var password = "neo"
            }
            let client = try BoltClient(CustomConfig())
            let result = client.connectSync()
            switch result {
            case .failure(_):
                return "Error while connecting"
            case .success(_):
                return "Connected to neo4j"
            }
        }
    }

But getting:

    [ NOTICE ] Server starting on http://127.0.0.1:8080
    [ INFO ] GET /
    get /
    Precondition failed: BUG DETECTED: wait() must not be called when on an EventLoop.
    Calling wait() on any EventLoop can lead to
      - deadlocks
      - stalling processing of other connections (Channels) that are handled on the EventLoop that wait was called on
    Further information:
      - current eventLoop: Optional(SelectableEventLoop { selector = Selector { descriptor = 7 }, scheduledTasks = PriorityQueue(count: 0): [ScheduledTask(readyTime: 369712427183210)] })
      - event loop associated to future: NIOTransportServices.NIOTSEventLoop: file /Users/Tieme/Projects/Neo4j-test/.build/checkouts/Bolt-swift/Sources/UnencryptedSocket.swift, line 87
    2019-11-30 14:12:31.494618+0100 Run[29034:15536002] Precondition failed: BUG DETECTED: wait() must not be called when on an EventLoop.
    Calling wait() on any EventLoop can lead to
      - deadlocks
      - stalling processing of other connections (Channels) that are handled on the EventLoop that wait was called on
    Further information:
      - current eventLoop: Optional(SelectableEventLoop { selector = Selector { descriptor = 7 }, scheduledTasks = PriorityQueue(count: 0): [ScheduledTask(readyTime: 369712427183210)] })
      - event loop associated to future: NIOTransportServices.NIOTSEventLoop: file /Users/Tieme/Projects/Neo4j-test/.build/checkouts/Bolt-swift/Sources/UnencryptedSocket.swift, line 87
    (edited)

It should be possible to use Neo4j under Vapor right? Any tips or working examples?

**Comment by teameh** [comment]
Apparently Vapor v4 does not play well with the latest Bolt version. I revived some old Neo4j and Bolt versions to be able to use Vapor v3 for now until this is fixed, feel free to use them while this is pending. 

```
.package(url: "https://github.com/teameh/Neo4j-swift.git", .branch("4.0.2-revived")),
.package(url: "https://github.com/teameh/Bolt-swift.git", .branch("1.0.2-swift-tools-4")),
```

**Comment by niklassaers** [comment]
I know the thread seems dead, so just wanted to give an update. I'm working on removing those wait()'s (would love help on that, btw), but when I want to test them with Vapor I've often gotten lost in the SPM resolver just seemingly waiting forever. If anyone wants to trade notes and help getting this working with Vapor 4 again, that'd be awesome :-)

**Comment by teameh** [comment]
👍 for working on this issue! 

I had  the same issue with SPM. I tried to find the issue/thread/question with some helpful tips but I can't find it any more.. I think the issue is related to the packager being unable to satisfy all requirements. Did you try to use the verbose flag `-v`? 

I did not find this link before but apparently there's a bug report for this as well https://bugs.swift.org/browse/SR-6389

Good luck!

**Comment by kiliankoe** [comment]
If the resolver is causing issues, try passing the flag `--enable-pubgrub-resolver` to SwiftPM. This uses the new (and completely different) resolver instead which will ideally just work, but if not should at least provide some more helpful diagnostics.

**Comment by niklassaers** [comment]
Wow, I should have asked for help so much earlier. :-) --enable-pubgrub-resolver got me out of the rut, and -v explained to me why. I'm back on track :-)

**Comment by teameh** [comment]
Ah yeah, that flag is very helpful. Found the article I couldn't fine before, thanks @kiliankoe 

https://forums.swift.org/t/infinite-execution-in-swift-build-while-resolving-dependencies/26637/3

@niklassaers good to hear! Good luck.

**Comment by niklassaers** [comment]
Update: the work is coming well along, tests are starting to get green, data coming through reliably. It'll be quite a bit of rework of [Bolt-Swift](https://github.com/Neo4j-Swift/Bolt-swift), so unsure how it'll affect the Theo API right now.

**Comment by niklassaers** [comment]
If you have the time, I'd love some review of [https://github.com/Neo4j-Swift/Bolt-swift/tree/dev/nonblocking](https://github.com/Neo4j-Swift/Bolt-swift/tree/dev/nonblocking)

**Comment by niklassaers** [comment]
(seems the API impact on Theo will be much smaller than it was on Bolt-Swift)

**Comment by teameh** [comment]
Cool that the work is coming along! I'm not qualified enough to review this though, sorry.

**Comment by ivoleko** [comment]
I do not understand why CocoaPods version of the Theo use Bolt 5.0.0 dependency, and Swift package manager use master branch of Bolt?
With Swift Package Manager, Theo crashes all the time on iOS. With CocoaPods, it works without any problem.

**Comment by timcolson** [comment]
NewB. Today, worked hard to get Vapor v4.10.0 running, learned about SPM. I'm a long time Neo4j fan, so it was the first additional dependency I added. Got the error that led me here. Just chiming in to let you know at least one more human on the planet is interested in Theo, Neo, and Vapor4. :)

**Comment by niklassaers** [comment]
Just a quick comment to let you know that I’m making good progress on this again after having been stuck on a bug in my code for a good while.

**Comment by niklassaers** [comment]
(and thanks for the pings in here, that’s very motivating too :-) )

**Comment by niklassaers** [comment]
I've just pushed an version to the [dev/non-blocking branch](https://github.com/Neo4j-Swift/Neo4j-Swift/tree/dev/non-blocking), together with the [dev/bolt-v3 branch of Bolt-swift](https://github.com/Neo4j-Swift/Bolt-swift/tree/dev/bolt-v3). I'll be moving it over to stable soon, with updated Podfiles, but if you have the chance it'd be awesome with feedback before then. :-)

**Comment by niklassaers** [comment]
Version 5.2.0 has been released, I'd love feedback on how these issues are now :-)

**Comment by IsmailAhmedIsmail** [comment]
Whoa, I've opened this issue out of passion for learning server-side swift in my spare time with my favorite Graph DB Neo4j.
Coming back a year later to discover this great progress from you. All I can say really is thank you for your support, even it seemed as a dead thread.
This, along with fix on dev/5.2.1 for compatibility issues and encryption, led it back to work successfully.

**Comment by niklassaers** [comment]
Thanks, Ismail, I’m very happy to hear it’s working well for you. :-)

>
